### PR TITLE
Issue 1129 julia

### LINKED ISSUE #1129: Generate Julia code from a pybamm expression tree
Mimic `EvaluatorPython` to generate a Julia function from an expression tree.
The first step is to generate a pure Julia function.
Later on, we can generate functions in the `ModelingToolkit.jl` format

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Some basic functionality for generating PyBaMM expression trees in Julia. The code is pretty horrible, but it's a start

Fixes #1129 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/expression_tree/operations/evaluate_julia.py [hunk 1]
```diff
@@ -0,0 +1,1101 @@
...
+            child_eval = child.evaluate()
+            if isinstance(child_eval, numbers.Number):
+                children_vars.append(str(child_eval))
+            else:
+                children_vars.append(id_to_julia_variable(child.id, "const"))
+        else:
+            children_vars.append(id_to_julia_variable(child.id, "cache"))
+
+    if isinstance(symbol, pybamm.BinaryOperator):
+        # Multiplication and Division need special handling for scipy sparse matrices
```
Comment (martinjrobins) [inline pybamm/expression_tree/operations/evaluate_julia.py:154]: is this comment current? I can't see any special handling for scipy sparse matrices

File: tests/unit/test_models/test_base_model_generate_julia_diffeq.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    Julia(compiled_modules=False)
+    from julia import Main
+
+    # load julia libraries required for evaluating the strings
+    Main.eval("using SparseArrays, LinearAlgebra")
+
+
+@unittest.skipIf(not have_julia, "Julia not installed")
+@unittest.skipIf(platform.system() == "Windows", "Skipped for Windows")
+class TestBaseModelGenerateJuliaDiffEq(unittest.TestCase):
```
Comment (martinjrobins) [inline tests/unit/test_models/test_base_model_generate_julia_diffeq.py:21]: looks like most of these tests are not using `Main`, so can we turn them on for windows?

File: tests/unit/test_solvers/test_julia_mtk.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+import pybamm
+
+import unittest
+from platform import system
+
+# import numpy as np
+
+# julia imports
+have_julia = pybamm.have_julia()
+# if have_julia and system() != "Windows":
```
Comment (martinjrobins) [inline tests/unit/test_solvers/test_julia_mtk.py:13]: remove commented code

**Review by martinjrobins — APPROVED** [review]
looks great, thanks @tinosulzer , some minor comments below

File: pybamm/util.py [hunk 4]
```diff
@@ -367,6 +367,21 @@ def get_parameters_filepath(path):
         return os.path.join(pybamm.__path__[0], path)
 
 
+def have_julia():
+    """
+    Checks whether the Julia programming language has been installed
+    """
+    import subprocess
```
Comment (Saransh-cpp) [inline pybamm/util.py:374]:
```suggestion
```
`subprocess` imported twice

File: setup.py [hunk 5]
```diff
@@ -186,6 +186,8 @@ def compile_KLU():
         "scikit-fem>=0.2.0",
         "casadi>=3.5.0",
         "imageio>=2.9.0",
+        "julia>=0.5.6",
+        # Can be installed even if julia is not installed
```
Comment (brosaplanella) [inline setup.py:190]: Does this line refer to the packages below or was there an additional julia package?

**Review by brosaplanella — APPROVED** [review]
Looks good, thanks Tino!

File: setup.py (hunk 5 above)
Comment (tinosulzer) [reply setup.py:190]: This refers to the julia package, the pip package can be installed even if the julia programming language is not installed (so there is no need to make the pip package an optional dependency). Edited the comment

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
 -   Added "Discharge energy [W.h]", which is the integral of the power in Watts, as an optional output. Set the option "calculate discharge energy" to "true" to get this output ("false" by default, since it can slow down some of the simple models) ([#1969](https://github.com/pybamm-team/PyBaMM/pull/1969)))
 -   Added an option "calculate heat source for isothermal models" to choose whether or not the heat generation terms are computed when running models with the option `thermal="isothermal"`  ([#1958](https://github.com/pybamm-team/PyBaMM/pull/1958))
+-   Added functionality to generate Julia expressions from a model. See [PyBaMM.jl](https://github.com/tinosulzer/PyBaMM.jl) for how to use these ([#1942](https://github.com/pybamm-team/PyBaMM/pull/1942)))
```
Comment (brosaplanella) [inline CHANGELOG.md:19]: This line needs to be moved to the unreleased section.

**Pull request merged**